test/functional/wallet_simulaterawtx.py

```diff
@@ -0,0 +1,129 @@
+        assert_equal(w1.getbalance(), 0.0)
+
+        address1 = w1.getnewaddress()
+        address2 = w1.getnewaddress()
+
+        # Add address1 as watch-only to w2
+        w2.importpubkey(pubkey=w1.getaddressinfo(address1)["pubkey"])
+
+        tx1 = node.createrawtransaction([], [{address1: 5.0}])
+        tx2 = node.createrawtransaction([], [{address2: 10.0}])
+
+        # w0 should be unaffected, w2 should see +5 for tx1
+        assert_equal(w0.simulaterawtransaction([tx1])["balance_change"], 0.0)
+        assert_equal(w2.simulaterawtransaction([tx1])["balance_change"], 5.0)
+
+        # w1 should see +5 balance for tx1
+        assert_equal(w1.simulaterawtransaction([tx1])["balance_change"], 5.0)
+
+        # w0 should be unaffected, w2 should see +5 for both transactions
+        assert_equal(w0.simulaterawtransaction([tx1, tx2])["balance_change"], 0.0)
+        assert_equal(w2.simulaterawtransaction([tx1, tx2])["balance_change"], 5.0)
+
+        # w1 should see +15 balance for both transactions
+        assert_equal(w1.simulaterawtransaction([tx1, tx2])["balance_change"], 15.0)
+
+        # w0 funds transaction; it should now see a decrease in (tx fee and payment), and w1 should see the same as above
+        funding = w0.fundrawtransaction(tx1)
+        tx1 = funding["hex"]
+        tx1changepos = funding["changepos"]
+        bitcoin_fee = Decimal(funding["fee"])
+
+        # w0 sees fee + 5 btc decrease, w2 sees + 5 btc
+        assert_approx(w0.simulaterawtransaction([tx1])["balance_change"], -(Decimal("5") + bitcoin_fee))
+        assert_approx(w2.simulaterawtransaction([tx1])["balance_change"], Decimal("5"))
+
+        # w1 sees same as before
+        assert_equal(w1.simulaterawtransaction([tx1])["balance_change"], 5.0)
+
+        # same inputs (tx) more than once should error
+        assert_raises_rpc_error(-8, "Transaction(s) are spending the same output more than once", w0.simulaterawtransaction, [tx1,tx1])
+
+        tx1ob = node.decoderawtransaction(tx1)
+        tx1hex = tx1ob["txid"]
+        tx1vout = 1 - tx1changepos
+        # tx3 spends new w1 UTXO paying to w0
+        tx3 = node.createrawtransaction([{"txid": tx1hex, "vout": tx1vout}], {w0.getnewaddress(): 4.9999})
+        # tx4 spends new w1 UTXO paying to w1
+        tx4 = node.createrawtransaction([{"txid": tx1hex, "vout": tx1vout}], {w1.getnewaddress(): 4.9999})
+
+        # on their own, both should fail due to missing input(s)
+        assert_raises_rpc_error(-8, "One or more transaction inputs have been spent already", w0.simulaterawtransaction, [tx3])
+        assert_raises_rpc_error(-8, "One or more transaction inputs have been spent already", w1.simulaterawtransaction, [tx3])
+        assert_raises_rpc_error(-8, "One or more transaction inputs have been spent already", w0.simulaterawtransaction, [tx4])
+        assert_raises_rpc_error(-8, "One or more transaction inputs have been spent already", w1.simulaterawtransaction, [tx4])
```

Comment:
Should these 4 cases raise the exception `One or more transaction inputs are missing` ?

Anyway, this exception is not being tested.

Reply:
You're right. I don't think that exception will ever be triggered, because `coins` is always populated with all the outputs in `mtx.vin` in the earlier block. I think the best approach is to simply reword this exception to say "spent or missing" and drop the other one.